src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -38101,5 +38102,120 @@ void M()
                 //             n = base.F.Length;
                 Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "base.F").WithLocation(21, 17));
         }
+
+        [WorkItem(29049, "https://github.com/dotnet/roslyn/issues/29049")]
+        [Fact]
+        public void ConstraintFromMetadata()
+        {
+            var source0 =
+@"using System;
+public class A<T> where T : IEquatable<string> { }";
+            var source =
+@"class B
+{
+    static void Main()
+    {
+        object o = new A<string?>(); // warning
+    }
+}";
+            var comp0 = CreateCompilation(source0, parseOptions: TestOptions.Regular8);
+            var ref0 = comp0.EmitToImageReference();
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8, references: new[] { ref0 });
+            var diagnostics = comp.GetDiagnostics();
+            diagnostics.Verify(
+                // (5,26): warning CS8631: The type 'string?' cannot be used as type parameter 'T' in the generic type or method 'A<T>'. Nullability of type argument 'string?' doesn't match constraint type 'System.IEquatable<string>'.
+                //         object o = new A<string?>(); // warning
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInTypeParameterConstraint, "string?").WithArguments("A<T>", "System.IEquatable<string>", "T", "string?").WithLocation(5, 26));
+            // Diagnostics must support GetHashCode() and Equals(), to allow removing
+            // duplicates (see CommonCompiler.ReportErrors.)
+            foreach (var diagnostic in diagnostics)
+            {
+                diagnostic.GetHashCode();
+                Assert.True(diagnostic.Equals(diagnostic));
+            }
+        }
+
+        [WorkItem(29041, "https://github.com/dotnet/roslyn/issues/29041")]
+        [WorkItem(29048, "https://github.com/dotnet/roslyn/issues/29048")]
+        [Fact]
+        public void ConstraintCyclesFromMetadata()
+        {
+            var source0 =
+@"using System;
+public class A0<T> where T : IEquatable<T> { }
+public class A1<T> where T : class, IEquatable<T> { }
+public class A2<T> where T : class, IEquatable<T?> { }
+public class A3<T> where T : struct, IEquatable<T> { }
+public class A4<T> where T : struct, IEquatable<T?> { }
+public class A5<T> where T : IEquatable<string?> { }
+public class A6<T> where T : IEquatable<int?> { }";
+            var source =
+@"class B
+{
+    static void Main()
+    {
+        new A0<string?>();
```

Comment:
>string? [](start = 15, length = 7)

From our discussion, I expect a warning on `A0<string?>` once the substitution rule is adjusted for oblivious/nullable mixing together.

Reply:
Added comment.

---
In reply to: [207677078](https://github.com/dotnet/roslyn/pull/29045#discussion_r207677078) [](ancestors = 207677078)